### DISCUSSION TITLE
Does iTwinUI work with Qwik apps?

### DISCUSSION THREAD

**Discussion by r100-stack** [iTwin/iTwinUI#934] [Q&A] [answered]
I was trying to add iTwinUI react components in a demo [Qwik](https://qwik.builder.io/) app. They say that we can include React UI libraries in Qwik apps using the [`qwikify` integration](https://qwik.builder.io/integrations/integration/react/).

However, when following the above guide, I was not able to add iTwinUI React components to the project. I constantly keep getting different errors. One of them is:

![image](https://user-images.githubusercontent.com/45748283/209479091-2aef6406-8f02-4258-b58a-aec1189e81c8.png)

After getting this error, I thought to import directly from `cjs` instead of the root.

```diff
- import { IconButton } from "@itwin/itwinui-react";
+ import { IconButton } from "@itwin/itwinui-react/cjs";
```

I then keep getting this almost identical error:

![image](https://user-images.githubusercontent.com/45748283/209479148-8905d4ae-2def-4f83-b6db-21c0564495e6.png)

I tried changing with node versions 16 and 18 but faced the same. I also tried changing the import paths to something more specific:

```tsx
import { IconButton } from "@itwin/itwinui-react/cjs/core/Buttons/IconButton";
```

But that didn't work too.

When trying `npm run preview`, it gives me a syntax error in `global.css`

![image](https://user-images.githubusercontent.com/45748283/209486077-ead941a8-9f5c-47a9-843e-dad46f2061ab.png)

Not sure if they are related, but I also received a syntax error when trying to use `@itwin/itwinui-react` in Next JS but without using [`next-transpile-modules`](https://www.npmjs.com/package/next-transpile-modules).

![image](https://user-images.githubusercontent.com/45748283/209486243-d3b30041-ad96-4111-b479-b859cdc61a5f.png)

Here is a sample repo I created where I tried using `iTwinUI-react`'s `IconButton`:
https://github.com/rohan-kadkol/demo-qwik-itwinui

Is there something I am missing? Or maybe is there a reason why iTwinUI components won't work in Qwik apps?

**Comment by mayank99** [comment, marked as answer]
This error likely has nothing to do with qwik itself.

I cannot access your sample repo (might be private) but if I had to guess, you are using qwik city, which is an SSR framework for qwik (similar to nextjs or astro). qwik city uses vite for SSR, and it's tricky to make it work out of the box.

Try adding `type: module` to your package.json and see if it changes the error you get.

After that, try adding `@itwin/itwinui-react` and `@itwin/itwinui-css` to [`ssr.noExternal`](https://vitejs.dev/config/ssr-options.html#ssr-noexternal) in your vite config.

```js
export default defineConfig({
  /* ...rest of your vite config */

  ssr: {
    noExternal: [
      '@itwin/itwinui-react',
      '@itwin/itwinui-css',
    ]
  }
});
```

**Reply by r100-stack** [reply]
Thanks for the idea. I had tried your suggestions a while ago but forgot to reply as I was trying other approaches based on your answers but didn't complete that.

When I tried it, changing to `type: module` or adding the iTwin packages to `ssr.noExternal` in the vite config didn't work. Maybe I did something wrong at that time, not sure. But I will retry again sometime today and let you know.

> I cannot access your sample repo (might be private)

Sorry, yes, it was private. It is public now.

**Reply by mayank99** [reply]
did you try `ssr.noExternal`?

and the repo is still private

**Reply by r100-stack** [reply]
> and the repo is still private

Ohh, not sure why that is. I tried making it public again. Does it work now?

**Reply by mayank99** [reply]
yes it does now

**Reply by mayank99** [reply]
can you narrow down the repro to just qwikcity+react+itwinui by removing mui and emotion?

**Reply by r100-stack** [reply]
I recloned the repo, tried out `ssr.noExternal`, and it worked! Thanks!

![image](https://user-images.githubusercontent.com/45748283/211862866-03b08572-c5bc-4f3f-bf37-b9ba108e7531.png)

I guess I might have done something wrong when I tried it before. Maybe some local uncommitted changes I made that caused their own errors that I confused for errors because of `ssr.noExternal`. Since I recloned the repo, those local changes were not there.

Althought the visual part works, `onClick` intereactions with it don't seem to work. Maybe I need to make some more Qwik specific changes. When I try it out, will let you know if I could get the interaction to work.
